### DISCUSSION TITLE
AsyncNullConnectionPool with Notify/Listen

### DISCUSSION THREAD

**Discussion by MnrBnm** [psycopg/psycopg#292] [Q&A]
Hi,

I want to be able to listen to a channel with my AsyncNullConnectionPool. In the documentation, there is no example for that. Let me explain a little bit what I'm trying to do. I have two python scripts running : 

- One Insert/delete new rows to my table ITEM 
- The other one is WAITING for new rows in my table ITEM to start doing some stuff. What i used to do is check my table ITEM each 10 sec to see if there is something new but this is not the best approach i believe. 

I know we can use pg_notify to send notification to all the listener but I can't manage to make it work. 

Here is a small snippet of my code and how i access my db : 

db.py
```
class Db:
   async def connection_db(self,..):
        uri = ".."
        #It's in a class
        self.pool = psycopg_pool.AsyncNullConnectionPool(uri, min_size=0, max_size=50)

    async def add_row_db(self, row):
        async with self.pool.connection() as conn:
            async with conn.cursor(row_factory=dict_row) as cur:
                await cur.execute("INSERT......", row)

```

add_rows_db.py
```
db = Db(..)
while True:
   #do stuff
   await db.add_row_db(row)
   ..
```

listener_rows.py
```
db = Db(..)
def work_on_row(..):
  # means a new row has been added 
   while True:
      #do stuff with row information...
```

I would like to have my function "work_on_row" (from file : listener.rows.py) called each time a new row is added (eventually the same thing with another function when we have a delete) 

thank you in advance !

**Comment by dvarrazzo** [comment]
Docs are available in [async notifications](https://www.psycopg.org/psycopg3/docs/advanced/async.html#asynchronous-notifications). Your listener should look like:

```python
    # the connection must be autocommit.
    conn = await psycopg.AsyncConnection.connect("...", autocommit=True)
    async with conn:
        await conn.execute("listen channel_name")
        async for notify in conn.notifies():
             process_record(int(notify.payload))
```

on the producer side:

```python
await cur.execute("INSERT ... RETURNING id", row)
id = (await cur.fetchone())[0]
await cur.execute("SELECT pg_notify('channel_name', %s::text)", [id])
```

Sending the notification can be done automatically if you register a trigger on that table, something like:

```
create or replace function my_table_insert() returns trigger
language plpgsql as $$
begin
    perform pg_notify('channel_name', new.id::text);
    return null;
end
$$;

create trigger my_table_insert
after insert on my_table
for each row execute procedure my_table_insert();
```
